frontend/frontend-filters.php

```diff
@@ -161,42 +166,141 @@ public function widget_display_callback( $instance ) {
+			unset( $sidebars_widgets[ $sidebar ][ $key ] );
 		}
+		return $sidebars_widgets;
+	}
 
+	/**
+	 * Browse the widgets sidebars and sort the ones that should be displayed or not.
+	 *
+	 * @since 3.1
+	 *
+	 * @param  array $sidebars_widgets       An associative array of sidebars and their widgets
+	 * @param  array $wp_registered_widgets  Array of all registered widgets.
+	 * @param  array $callback               Array of the callback function to call.
+	 * @return array                         An associative array of sidebars and their widgets
+	 */
+	public function sort_widgets_sidebars( $sidebars_widgets, $wp_registered_widgets, $callback ) {
 		foreach ( $sidebars_widgets as $sidebar => $widgets ) {
 			if ( 'wp_inactive_widgets' === $sidebar || empty( $widgets ) ) {
 				continue;
 			}
 
 			foreach ( $widgets as $key => $widget ) {
-				// Nothing can be done if the widget is created using pre WP2.8 API :(
-				// There is no object, so we can't access it to get the widget options
-				if ( ! isset( $wp_registered_widgets[ $widget ]['callback'] ) || ! is_array( $wp_registered_widgets[ $widget ]['callback'] ) || ! isset( $wp_registered_widgets[ $widget ]['callback'][0] ) || ! is_object( $wp_registered_widgets[ $widget ]['callback'][0] ) || ! method_exists( $wp_registered_widgets[ $widget ]['callback'][0], 'get_settings' ) ) {
+				if ( ! $this->is_widget_object( $wp_registered_widgets, $widget ) ) {
 					continue;
 				}
 
-				$widget_settings = $wp_registered_widgets[ $widget ]['callback'][0]->get_settings();
-				$number          = $wp_registered_widgets[ $widget ]['params'][0]['number'];
+				$widget_data = $this->get_widget_data( $wp_registered_widgets, $widget );
 
-				// Remove the widget if not visible in the current language
-				if ( ! empty( $widget_settings[ $number ]['pll_lang'] ) && $widget_settings[ $number ]['pll_lang'] !== $this->curlang->slug ) {
-					unset( $sidebars_widgets[ $sidebar ][ $key ] );
-				}
+				$args = array(
+					$widget_data,
+					$sidebars_widgets,
+					$sidebar,
+					$key,
+				);
+				$sidebars_widgets = call_user_func_array( $callback, $args );
 			}
 		}
+		return $sidebars_widgets;
+	}
+
+	/**
+	 * Init widgets sidebars, checks if they already exist in the cache to use them later.
+	 *
+	 * @since 3.1
+	 *
+	 * @param array $sidebars_widgets An associative array of sidebars and their widgets
+	 * @return mixed|void
+	 */
+	public function init_sidebars_widgets( $sidebars_widgets ) {
+		if ( empty( $wp_registered_widgets ) ) {
+			return $sidebars_widgets;
+		}
 
-		$this->cache->set( "sidebars_widgets_{$cache_key}", $sidebars_widgets );
+		$this->cache_key         = md5( maybe_serialize( $sidebars_widgets ) );
+		$_sidebars_widgets = $this->cache->get( "sidebars_widgets_{$this->cache_key}" );
 
-		return $sidebars_widgets;
+		if ( false !== $_sidebars_widgets ) {
+			return $_sidebars_widgets;
+		}
```

Comment:
Now the method has no default return value, which could be troublesome for filters set by others plugins on the same hook.

Reply:
See the following commits.

Reply:
Since we are supposed to receive an array, maybe we should return an array as well? That way, plugins that hook up later on will not risk to break because of an incorrect type.

Reply:
This part of the code was supposed to quit the method : https://github.com/polylang/polylang/blob/master/frontend/frontend-filters.php#L164-L173